dll/win32/iphlpapi/icmp.c

```diff
@@ -1,1129 +1,575 @@
-    LPVOID                   RequestData,
-    WORD                     RequestSize,
-    PIP_OPTION_INFORMATION   RequestOptions,
-    LPVOID                   ReplyBuffer,
-    DWORD                    ReplySize,
-    DWORD                    Timeout
-    )
-{
-    icmp_t* icp=(icmp_t*)IcmpHandle;
-    unsigned char* reqbuf;
-    int reqsize;
-
-    struct icmp_echo_reply* ier;
-    struct ip* ip_header;
-    struct icmp* icmp_header;
-    char* endbuf;
-    int ip_header_len;
-    int maxlen;
-#ifdef __REACTOS__
-    fd_set fdr;
-    struct timeval timeout;
-#else
-    struct pollfd fdr;
-#endif
-    DWORD send_time,recv_time;
-    struct sockaddr_in addr;
-    socklen_t addrlen;
-    unsigned short id,seq,cksum;
-    int res;
-
-    if (IcmpHandle==INVALID_HANDLE_VALUE) {
-        /* FIXME: in fact win98 seems to ignore the handle value !!! */
-        SetLastError(ERROR_INVALID_HANDLE);
-        return 0;
-    }
 
-#ifdef __REACTOS__
-    if (ReplySize<sizeof(ICMP_ECHO_REPLY)) {
-        SetLastError(ERROR_INVALID_PARAMETER);
-        return 0;
-    }
+    CloseHandle(hEvent);
+    HeapFree(GetProcessHeap(), 0, InputBuffer);
 
-    if (ReplySize-RequestSize<sizeof(ICMP_ECHO_REPLY)) {
-        SetLastError(IP_GENERAL_FAILURE);
+    if (!NT_SUCCESS(Status))
+    {
+        SetLastError(RtlNtStatusToDosError(Status));
         return 0;
     }
 
-    if (!ReplyBuffer) {
-        SetLastError(ERROR_INVALID_PARAMETER);
-        return 0;
+    Status = ((PICMP_ECHO_REPLY)ReplyBuffer)->Status;
+    if (Status != IP_SUCCESS)
+    {
+        SetLastError(Status);
     }
 
-    if (Timeout == 0 || Timeout == -1) {
-        SetLastError(ERROR_INVALID_PARAMETER);
-#else
-    if (ReplySize<sizeof(ICMP_ECHO_REPLY)+ICMP_MINLEN) {
-        SetLastError(IP_BUF_TOO_SMALL);
-#endif
-        return 0;
-    }
-    /* check the request size against SO_MAX_MSG_SIZE using getsockopt */
+    nReplies = ((PICMP_ECHO_REPLY)ReplyBuffer)->Reserved;
+    ((PICMP_ECHO_REPLY)ReplyBuffer)->Reserved = 0;
 
-    if (!DestinationAddress) {
-        SetLastError(ERROR_INVALID_NETNAME);
-        return 0;
-    }
+    return nReplies;
+}
 
-#ifndef __REACTOS__
-    if (icp->sid < 0) {
-        WARN("using system ping command since SOCK_RAW was not supported.\n");
-        return system_icmp(DestinationAddress, RequestData, RequestSize,
-                           RequestOptions, ReplyBuffer, ReplySize, Timeout);
-    }
-#endif
-
-    /* Prepare the request */
-#ifdef __REACTOS__
-    id = GetCurrentProcessId() & 0xFFFF;
-#else
-    id=getpid() & 0xFFFF;
-#endif
-    seq=InterlockedIncrement(&icmp_sequence) & 0xFFFF;
-
-#ifdef __REACTOS__
-    reqsize=ICMP_MINLEN;
-    if (RequestData && RequestSize > 0)
-        reqsize += RequestSize;
-#else
-    reqsize=ICMP_MINLEN+RequestSize;
-#endif
-    reqbuf=HeapAlloc(GetProcessHeap(), 0, reqsize);
-    if (reqbuf==NULL) {
-        SetLastError(ERROR_OUTOFMEMORY);
+/**
+ * IcmpSendEcho
+ * @implemented
+ */
+DWORD
+WINAPI
+IcmpSendEcho(
```

Comment:
@crawfxrd Is there any reason why `IcmpSendEcho` does not just call `IcmpSendEcho2` and has almost identical code?

Reply:
I implemented each function independently in tandem with the tests to ensure I had the correct behavior as I implemented them. I just never followed up with refactoring it.

Reply:
From my tests, `ReplySize` is checked identically (`ReplySize < sizeof(ICMP_REPLY_SIZE) == 28)`) At least on w2k3

Reply:
I don't know why Microsoft made is so the space is _always_ required. The only case I can think of it being useful is if `ApcRoutine` is passed as `FARPROC` instead of`PIO_APC_ROUTINE`. (That's assuming the `IO_STATUS_BLOCK`s are even identical. Copies of the one passed in to `NtDeviceIoControlFile()`?)